### PR TITLE
Console Commands Optimization

### PR DESCRIPTION
**Pull request by kirmorozov**
### Description (*)

Put some resource-intensive constructors behind Proxy class.

### Fixed Issues (if relevant)

1. Fixes magento/magento2#<issue_number>

### Manual testing scenarios (*)

1. bin/magento
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by kirmorozov** [comment]
@magento run all tests

**Comment by kirmorozov** [comment]
@magento run `Functional Tests CE`,`Semantic Version Checker`

**Comment by kirmorozov** [comment]
@magento run Semantic Version Checker

**Comment by kirmorozov** [comment]
@magento run Functional Tests EE,Functional Tests CE,Functional Tests B2B

File: app/code/Magento/MediaStorage/etc/di.xml [hunk 1]
```diff
@@ -28,7 +28,11 @@
     </type>
     <type name="Magento\MediaStorage\Console\Command\ImagesResizeCommand">
         <arguments>
+            <argument name="appState" xsi:type="object">Magento\Framework\App\State</argument>
```
Comment (ihor-sviziev) [inline app/code/Magento/MediaStorage/etc/di.xml:31]: Should it have a Proxy suffix as well? If no - please remove it

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Hi @kirmorozov,
In general, your changes look good in terms of the 2.4-develop branch. However, I have one small comment, could you review and fix it?


For the 2.5-develop branch, we already have an amazing PR https://github.com/magento/magento2/pull/29355 that is supposed to permanently fix this issue for all places at once, so such changes should be probably removed.

**Comment by gabrieldagama** [comment]
Hi @kirmorozov, thanks for contributing! Can you provide some extra information on execution time improvement? Ideally, we need to see the difference between your solution and vanilla Magento.

Can you provide that information?

Thanks!

**Comment by ihor-sviziev** [comment]
Hi @gabrieldagama,
The idea behind this PR is not to create instances of the commands with all their dependencies (unfortunately, some classes have quite heavy constructors that that slowing down all the CLIs) until needed. That's what proxies do for us.

As I said above - the permanent fix will be implementing CLI command loader, but for 2.4.x it's not suitable https://github.com/magento/magento2/pull/29355

**Comment by gabrieldagama** [comment]
Hi @ihor-sviziev, we were discussing this PR at the Public Community Triage meeting. We understand the goal with applying proxies to the commands, but as you mentioned there, it probably doesn't make sense to accept such changes now and revert them when 2.5 gets released, mostly because it will cause BiC one more time.

We decided to close this PR for now in favor of the CommandLoader implementation on #29355.

@kirmorozov Thanks for contributing!

**Comment by kirmorozov** [comment]
@gabrieldagama you are referencing something in future, that does not work now, and maybe will never work.

**Comment by gabrieldagama** [comment]
Hi @kirmorozov,

Yes, you are right, the PR #29355 is aimed at `2.5` scope instead. Replacing the `CommandList` with `CommandLoader` may be a challenge but it seems a more definitive solution. What we were discussing at the Public Community Triage meeting is that bringing a temporary solution for a noncritical problem and having to revert it on the next minor release is not recommended.

Hope that makes sense, thanks again.

**Pull request closed without merging**